x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilder.java

```diff
@@ -0,0 +1,247 @@
+
+public class WeightedTokensQueryBuilder extends AbstractQueryBuilder<WeightedTokensQueryBuilder> {
+    public static final String NAME = "weighted_tokens";
+
+    public static final ParseField TOKENS_FIELD = new ParseField("tokens");
+    private final String fieldName;
+    private final Set<WeightedToken> tokens;
+    private final WeightedTokensThreshold threshold;
+
+    public WeightedTokensQueryBuilder(String fieldName, Set<WeightedToken> tokens) {
+        this(fieldName, tokens, null);
+    }
+
+    public WeightedTokensQueryBuilder(String fieldName, Set<WeightedToken> tokens, @Nullable WeightedTokensThreshold threshold) {
+        this.fieldName = Objects.requireNonNull(fieldName, "[" + NAME + "] requires a fieldName");
+        this.tokens = Objects.requireNonNull(tokens, "[" + NAME + "] requires tokens");
+        if (tokens.isEmpty()) {
+            throw new IllegalArgumentException("[" + NAME + "] requires at least one token");
+        }
+        this.threshold = Objects.requireNonNullElse(threshold, new WeightedTokensThreshold());
+    }
+
+    public WeightedTokensQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.tokens = in.readCollectionAsSet(WeightedToken::new);
+        this.threshold = in.readOptionalWriteable(WeightedTokensThreshold::new);
+    }
+
+    public String getFieldName() {
+        return fieldName;
+    }
+
+    @Nullable
+    public WeightedTokensThreshold getThreshold() {
+        return threshold;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeCollection(tokens);
+        out.writeOptionalWriteable(threshold);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.startObject(fieldName);
+        builder.startObject(TOKENS_FIELD.getPreferredName());
+        for (var token : tokens) {
+            token.toXContent(builder, params);
+        }
+        builder.endObject();
+        if (threshold != null) {
+            threshold.toXContent(builder, params);
+        }
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+        builder.endObject();
+    }
+
+    private float getAverageTokenFreqRatio(IndexReader reader, int fieldDocCount) throws IOException {
+        int numUniqueTokens = 0;
+        for (var leaf : reader.getContext().leaves()) {
+            var terms = leaf.reader().terms(fieldName);
+            if (terms != null) {
+                numUniqueTokens = (int) Math.max(terms.size(), numUniqueTokens);
+            }
+        }
+        if (numUniqueTokens == 0) {
+            return 0;
+        }
+        return (float) reader.getSumDocFreq(fieldName) / fieldDocCount / numUniqueTokens;
```

Comment:
So I'm not 100% sure how to interpret the terms here. I find the `Freq` confusing (since this usually means a fraction). Just to be clear on how I was expecting the average token frequency to be calculated:

Let the total number of tokens for the field in the index be T (i.e. the count of all tokens in every document)
Let the total number of documents - for which the field is populated - in the index be D
Let the total number of unique tokens be U

If the tokens are uniformly distributed, then each token occurs T / U times. Each token occurs at most once per document, so in this case each token occurs in a fraction of documents T / U / D. This is the fraction of documents we expect to contain each token on average if they are not over expressed in the corpus. (So this defines their expected information content, i.e. log(P(x)), with tokens occurring more frequently than this having lower than expected information content.) 

So by this reckoning `reader.getSumDocFreq(fieldName)` returns the total count of tokens in the index. Is this right?

Another observation is that I used the model vocabulary size (U = 30522) rather than the expressed token count. I can see arguments for both choices, but if anything I think the choice here is better. However, I just want to note that this was how my testing was done so we might want to revisit the default parameter choices I provided given this change.

I don't propose any changes for this PR, but would like to track somewhere that we need to evaluate with this choice. Since I need to do a run to evaluate no reranking across BEIR I can do this at the same time.

Reply:
I also think it would be worth adding a comment in this code to explain this calculation.

Reply:
> So by this reckoning reader.getSumDocFreq(fieldName) returns the total count of tokens in the index. Is this right?

That seems right yes,  `reader.getSumDocFreq(fieldName)` returns the sum of `TermsEnum.docFreq(Term)` for all terms in this field and `TermsEnum.docFreq(Term)` returns the number of documents containing the term.